### PR TITLE
tee-supplicant: fix plugin loading logic

### PR DESCRIPTION
**Pull request by kubiko**
List of the plugins to be loaded by tee-supplicant is constructed based on the content of the `TEE_PLUGIN_LOAD_PATH` directory. However when plugins are loaded only file name is passed to the `dlopen()` function.
This means that actual plugin is loaded at first hit based on `LD_LIBRARY_PATH` search path.
This could result in loading of completely different file than actually intended.

Solution is to pass full file path to the `dlopen()`, that is `TEE_PLUGIN_LOAD_PATH` + plugin filename.

### CODE REVIEW

**Comment by kubiko** [comment]
Rebasing on maser after #299 was merged in

**Comment by anisyanka** [comment]
```
Reviewed-by:  Aleksandr Anisimov <anisimov.alexander.s@gmail.com>
```

File: tee-supplicant/src/plugin.c [hunk 1]
```diff
@@ -136,7 +137,8 @@ TEEC_Result plugin_load_all(void)
 			return TEEC_ERROR_OUT_OF_MEMORY;
 		}
 
-		res = load_plugin((const char *)entry->d_name, p);
+		snprintf(path, sizeof(path), "%s/%s", supplicant_params.plugin_load_path, (const char *)entry->d_name);
```
Comment (jforissier) [inline tee-supplicant/src/plugin.c:140]: The cast is not needed I think.

**Review by jforissier — COMMENTED** [review]
One minor comment (please check), in any case:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: tee-supplicant/src/plugin.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kubiko) [reply tee-supplicant/src/plugin.c]: error handling added

File: tee-supplicant/src/plugin.c [hunk 2]
```diff
@@ -135,8 +137,15 @@ TEEC_Result plugin_load_all(void)
 			closedir(dir);
 			return TEEC_ERROR_OUT_OF_MEMORY;
 		}
-
-		res = load_plugin((const char *)entry->d_name, p);
+		r = snprintf(path, sizeof(path), "%s/%s",
+		             supplicant_params.plugin_load_path, entry->d_name);
+		if (r < 0 || r == sizeof(path)) {
```
Comment (jforissier) [inline tee-supplicant/src/plugin.c:142]: `r >= sizeof(path)`

**Comment by jenswi-linaro** [comment]
With @jforissier's comment addressed:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by kubiko** [comment]
> With @jforissier's comment addressed:
> `Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

added

File: tee-supplicant/src/plugin.c (hunk 2 above)
Comment (kubiko) [reply tee-supplicant/src/plugin.c]: fixed

**Comment by jforissier** [comment]
```
 19252:  src/plugin.c:142:18: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
 19253:    142 |   if (r < 0 || r >= sizeof(path)) {
 19254:        |                  ^~
 19255:  cc1: all warnings being treated as errors
```
This should be `if (r < 0 || r >= (int)sizeof(path))` I suppose.

**Comment by kubiko** [comment]
> ```
>  19252:  src/plugin.c:142:18: error: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Werror=sign-compare]
>  19253:    142 |   if (r < 0 || r >= sizeof(path)) {
>  19254:        |                  ^~
>  19255:  cc1: all warnings being treated as errors
> ```
> 
> 
>     
>  
>   
> 
> This should be `if (r < 0 || r >= (int)sizeof(path))` I suppose.

thank you for that.
I run test build here and that seems to pass

**Comment by jforissier** [comment]
>  I run test build here and that seems to pass

Not surprising, CI builds with many warnings enabled and with `-Werror`. Thanks for the update.

**Pull request merged**